### PR TITLE
Add switch to save into empty file

### LINKED ISSUE #1144: Inconsistent handling of password-protected PDFs 1.11.x vs 1.10.0
Describe the bug
----------------

Up until version 1.10.0 we used PDF Arranger to decrypt encrypted PDF documents. 
Open encrypted file with password -> Export file / Save as = Exported PDF is not encrypted 

Since version 1.11.0, PDF Arranger saves PDF files with the password of the first imported/opened file if the first imported file was encrypted. I did not see this mentioned in the release notes. This “magical” behavior may be useful when working with only one document, but it can be confusing when working with different documents, wanting to save unencrypted documents or wanting to change the password / PDF restrictions of a PDF document, as there are no settings to configure the export settings.

To Reproduce
------------
Please always restart PDF Arranger 1.11.1 prior to testing any of the following scenarios: 
-> If you import an **unencrypted file first** the exported documents are not encrypted even if you import encrypted documents later.
-> If you import two encrypted documents, the **password of the document imported first is used for the export,** even if you change the order of the pages so that the pages of the document imported later appear first.
-> If you import a unencrypted PDF document first and encrypted documents later the export is always not encrypted
-> If you add a blank page to an empty PDF Arranger window, delete it and then import an encrypted PDF the export has no encryption 
-> Open document that is not encrypted and delete all pages - import an encrypted document in the now empty PDF Arranger window. Exported document is encrypted and has the password of the first document you opened  

Expected behavior
-----------------

Same as 1.10.x: Saved PDFs are not password protected.

Future feature: Add password settings (and maybe file restrictions) to document settings / options menu (or at least an option "Keep encryption / document restrictions of first imported document) 

System and Versions
-------------------

-   PDF Arranger Portable Windows versions 1.10.0 + 1.11.0 + 1.11.1 (Github releases)
-   Windows 10

### PR DESCRIPTION
**Pull request by StefRe**
Let user decide whether document-level information, such as outlines, tags, encryption passwords, etc., shall be taken from the first opened file and be preserved in the output file when saving or exporting to a single file. This became the default behavior with PDF Arranger 1.11 using pikepdf version 8 or newer, which uses QPDF jobs. Prior to that, the output was always built starting with a new, empty pdf document and thus all document-level information from the input file was discarded in the output. This old behavior can now be enforced by deactivating the check button "Retain document-level information of the first opened file" in the preferences dialog.
This new check button is only available when working with pikepdf 8 or newer.

Closes #1144

### CODE REVIEW

File: tests/test.py [hunk 1]
```diff
@@ -898,3 +895,60 @@
...
+            self.assertNotIn(warning, stdout)
+        else:
+            self.assertIn(warning, stdout)
+
+    def test_02_legacy(self):
+        """Test legacy setting"""
+        if have_pikepdf8():
+            self._mainmenu("Preferences")
+            dialog = self._app().child(roleName="dialog")
+            panels = self._find_by_role("panel", dialog)
```
[the remark this answers is not in the document]
Comment (StefRe) [reply tests/test.py:942]: fixed

File: pdfarranger/config.py [hunk 2]
```diff
@@ -299,6 +319,8 @@
             num2 = combo2.get_active()
             theme = themes[num2] if num2 != 0 else ""
             self.set_theme(theme)
+            if self.has_pikepdf8:
+                self.set_start_with_empty(not cb_retain.get_active())
```
[the remark this answers is not in the document]
Comment (StefRe) [reply pdfarranger/config.py:323]: is false positive, fixed by checking if variable exists

File: pdfarranger/config.py [hunk 3]
```diff
@@ -299,6 +319,10 @@
             num2 = combo2.get_active()
             theme = themes[num2] if num2 != 0 else ""
             self.set_theme(theme)
+            if self.has_pikepdf8:
+                # prevent CodeQL false positive "uninitialized local variable"
+                if cb_retain in vars():
```
[the remark this answers is not in the document]
Comment (StefRe) [reply pdfarranger/config.py]: fixed by defining it (anyways should have been `'cb_retain' in vars()` instead of `cb_retain in vars()`)

**Comment by kbengs** [comment]
`start_with_empty` does not sound perfect to me. What about preserve.. something or retain.. something ? Other than that I think it looks good. Are you able to fix the tests?

**Comment by StefRe** [comment]
> `start_with_empty` does not sound perfect to me

I know (especially because it's the negation of the label text) but it is the more technical internal description of what is going on (we start with an empty pdf and add all the pages to it). The user is not supposed to directly edit the config file, so I thought the name might be OK nonetheless.
As you have seen, I didn't use the job interface (option `--empty`) for pikepdf >= 8 as it turned out that it yields slightly different results for files with annotations. It didn't, however, investigate why exactly.

> Are you able to fix the tests?

I thinks so. The problem is that I can't run the dogtail tests locally as I'm on Arch Linux with at-spi2 version 2.54.0 and they renamed the push button role in 2.53.1 which breaks dogtail. I opened an [issue](https://gitlab.com/dogtail/dogtail/-/issues/35) and the dogtail people already came up with a fix which I'll test in the next couple of days so that I can fix the tests locally.

**Comment by StefRe** [comment]
The Snap build error seems to me to be unrelated to the PR. Could you have a look, please. I have no knowledge of snap.

**Comment by kbengs** [comment]
I think I have seen the snap fail sometimes but then work the next time again.

There are issues if you try to save a pdf which produces warnings, for example this:
https://drivecentre.ca/wp-content/uploads/2017/03/Vacon-NXP-NXC-User-Manual-DPD00890B-UK.pdf

**Review by kbengs — CHANGES_REQUESTED** [review]
If a PDF produces warnings at saving you will see exceptions in terminal and no warning window.

**Comment by StefRe** [comment]
> If a PDF produces warnings at saving you will see exceptions in terminal and no warning window.

Hm, I'm not sure what this is supposed to refer to? This PR doesn't change anything about error handling.

**Comment by kbengs** [comment]
wrapper() (line 245 exporter.py) is expecting export_msg (=warnings and errors at saving) as last argument. You have changed that to self.config. Becuase of that the export process cant pass messages to main process. You should try to open and save the file I linked a few post above. Try it before your commit and then after. If you have clicked to not show warnings on saving you need to edit config.ini and set "show-save-warnings = True"

**Comment by StefRe** [comment]
@kbengs I fixed the `export_msg` issue and tested it manually, but I'm not sure if I'll have the time to write a test for `show-save-warnings` within the next few days. If you like you can merge it without the test and I'll make a separate PR for the test. If not I'll mark this PR as draft and see if I can finish it next week.

**Review by kbengs — APPROVED** [review]

**Comment by kbengs** [comment]
I think it looks good. Do you agree @dreua ?

**Comment by StefRe** [comment]
Converting to draft until #1166 goes in.

File: tests/test.py [hunk 4]
```diff
@@ -924,3 +931,89 @@ def test_01_cli_version(self):
         self.assertIn("libqpdf", stdout)
         self.assertIn("OS-", stdout)
         self.assertIn("OS_Version-", stdout)
+
+
+class TestBatch10(PdfArrangerTest):
+    """Test show-save-warnings config settings"""
```
Comment (kbengs) [inline tests/test.py:937]: You are not testing show-save-warnings here, are you?

File: tests/test.py [hunk 5]
```diff
@@ -924,3 +931,89 @@ def test_01_cli_version(self):
...
+            self.assertTrue(checkboxes[0].checked)
+        else:
+            panels = self._find_by_role("panel", dialog)
+            self.assertNotIn(savepanel_name, [p.name for p in panels])
+        dialog.child(name="OK").click()
+        self._wait_cond(lambda: dialog.dead)
+
+        self._mainmenu("Save As…")
+        self._save_as_chooser("encrypted.pdf", ["encrypted.pdf"] )
+        with pikepdf.open(os.path.join(self.__class__.tmp, "encrypted.pdf"), password="foobar"):
```
Comment (kbengs) [inline tests/test.py:963]: Why is this "with pikepdf.." here, is it a mistake?

File: tests/test.py [hunk 6]
```diff
@@ -924,3 +931,89 @@ def test_01_cli_version(self):
...
+        with pikepdf.open(os.path.join(self.__class__.tmp, "unencrypted.pdf")):
+            # there should be no pikepdf._core.PasswordError
+            pass
+
+    def test_03_quit(self):
+        self._quit()
+
+
+class TestBatch11(PdfArrangerTest):
+    """Test metadata"""
```
Comment (kbengs) [inline tests/test.py:999]: Here you are testing show-save-warnings ?

File: tests/test.py (hunk 4 above)
Comment (StefRe) [reply tests/test.py:937]: Yes, thanks. This used to be `"""Test saving config settings"""` but I thought this is ambiguous (test the config settings for saving vs. test that config settings are correctly saved), so I decided to re-word it but got myself confused.  This should read 
```
class TestBatch10(PdfArrangerTest):
    """Test start-with-empty config settings""" 
```
which is easier to relate to the actual code. Will fix it.

File: tests/test.py (hunk 6 above)
Comment (StefRe) [reply tests/test.py:999]: Yes, thanks for spotting this mistake. This should read
```
class TestBatch11(PdfArrangerTest):
    """Test show-save-warnings config setting"""
```
This TestBatch11 will later also be used in # 1169 for metadata testing, but I'll amend the docstring then. 
Will fix it.

File: tests/test.py (hunk 5 above)
Comment (StefRe) [reply tests/test.py:963]: Yes - another good find by you :). Lines 963 and 964 should be deleted. Will fix it.

**Comment by StefRe** [comment]
@kbengs Thank you again very much for the careful reading of this PR. Next time I'll try to cause less confusion by doing one thing after the other and not cherry pick from my local development branch with different issues fixed at the same time that depend on each other.

**Comment by kbengs** [comment]
Thanks!

**Pull request merged**